delta-lake/delta-23x/src/main/scala/org/apache/spark/sql/delta/rapids/delta23x/GpuMergeIntoCommand.scala

```diff
@@ -0,0 +1,1330 @@
+import org.apache.spark.sql.execution.metric.{SQLMetric, SQLMetrics}
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types.{DataTypes, LongType, StringType, StructType}
+
+case class GpuMergeDataSizes(
+    @JsonDeserialize(contentAs = classOf[java.lang.Long])
+    rows: Option[Long] = None,
+    @JsonDeserialize(contentAs = classOf[java.lang.Long])
+    files: Option[Long] = None,
+    @JsonDeserialize(contentAs = classOf[java.lang.Long])
+    bytes: Option[Long] = None,
+    @JsonDeserialize(contentAs = classOf[java.lang.Long])
+    partitions: Option[Long] = None)
+
+/**
+ * Represents the state of a single merge clause:
+ * - merge clause's (optional) predicate
+ * - action type (insert, update, delete)
+ * - action's expressions
+ */
+case class GpuMergeClauseStats(
+    condition: Option[String],
+    actionType: String,
+    actionExpr: Seq[String])
+
+object GpuMergeClauseStats {
+  def apply(mergeClause: DeltaMergeIntoClause): GpuMergeClauseStats = {
+    GpuMergeClauseStats(
+      condition = mergeClause.condition.map(_.sql),
+      mergeClause.clauseType.toLowerCase(),
+      actionExpr = mergeClause.actions.map(_.sql))
+  }
+}
+
+/** State for a GPU merge operation */
+case class GpuMergeStats(
+    // Merge condition expression
+    conditionExpr: String,
+
+    // Expressions used in old MERGE stats, now always Null
+    updateConditionExpr: String,
+    updateExprs: Seq[String],
+    insertConditionExpr: String,
+    insertExprs: Seq[String],
+    deleteConditionExpr: String,
+
+    // Newer expressions used in MERGE with any number of MATCHED/NOT MATCHED/NOT MATCHED BY SOURCE
+    matchedStats: Seq[GpuMergeClauseStats],
+    notMatchedStats: Seq[GpuMergeClauseStats],
+    notMatchedBySourceStats: Seq[GpuMergeClauseStats],
+
+    // Timings
+    executionTimeMs: Long,
+    scanTimeMs: Long,
+    rewriteTimeMs: Long,
+
+    // Data sizes of source and target at different stages of processing
+    source: GpuMergeDataSizes,
+    targetBeforeSkipping: GpuMergeDataSizes,
+    targetAfterSkipping: GpuMergeDataSizes,
+    @JsonDeserialize(contentAs = classOf[java.lang.Long])
+    sourceRowsInSecondScan: Option[Long],
+
+    // Data change sizes
+    targetFilesRemoved: Long,
+    targetFilesAdded: Long,
+    @JsonDeserialize(contentAs = classOf[java.lang.Long])
+    targetChangeFilesAdded: Option[Long],
+    @JsonDeserialize(contentAs = classOf[java.lang.Long])
+    targetChangeFileBytes: Option[Long],
+    @JsonDeserialize(contentAs = classOf[java.lang.Long])
+    targetBytesRemoved: Option[Long],
+    @JsonDeserialize(contentAs = classOf[java.lang.Long])
+    targetBytesAdded: Option[Long],
+    @JsonDeserialize(contentAs = classOf[java.lang.Long])
+    targetPartitionsRemovedFrom: Option[Long],
+    @JsonDeserialize(contentAs = classOf[java.lang.Long])
+    targetPartitionsAddedTo: Option[Long],
+    targetRowsCopied: Long,
+    targetRowsUpdated: Long,
+    targetRowsMatchedUpdated: Long,
```

Comment:
I see that these are not in 2.4 of our plugin. Is that a bug?

Reply:
Filed #9581 to address this and the missing equals/hashCode override pointed out above in the Delta Lake 2.4 support.